distributed/worker_state_machine.py

```diff
@@ -1986,83 +1986,102 @@ def _transition_executing_error(
+                self._transition_from_executing(ts),
+                ({ts: "fetch"}, []),
+            )
         else:
-            # Flight/executing finished unsuccessfully, i.e. not in memory
-            assert finish != "memory"
-            next_state = ts.next
-            assert next_state in {"waiting", "fetch"}, next_state
-            assert ts.previous in {"executing", "long-running", "flight"}, ts.previous
-
-            if ts.previous in ("executing", "long-running"):
-                self._release_resources(ts)
-                self.executing.discard(ts)
-                self.long_running.discard(ts)
-
-            if next_state != finish:
-                recs, instructions = self._transition_generic_released(
-                    ts, stimulus_id=stimulus_id
-                )
-            recs[ts] = next_state
+            assert ts.previous == "flight"
+            assert ts.next == "waiting"
+            self.in_flight_tasks.discard(ts)
+            ts.coming_from = None
+            recs, instr = {ts: "waiting"}, []
 
-        return recs, instructions
+        ts.state = "released"
+        ts.done = False
+        ts.previous = None
+        ts.next = None
+        return recs, instr
 
     def _transition_resumed_fetch(
         self, ts: TaskState, *, stimulus_id: str
     ) -> RecsInstrs:
-        """See Worker._transition_from_resumed"""
-        recs, instructions = self._transition_from_resumed(
-            ts, "fetch", stimulus_id=stimulus_id
-        )
-        if self.validate:
-            # This would only be possible in a fetch->cancelled->resumed->fetch loop,
-            # but there are no transitions from fetch which set the state to cancelled.
-            # If this assertion failed, we' need to call _ensure_communicating like in
-            # the other transitions that set ts.status = "fetch".
-            assert ts.state != "fetch"
-        return recs, instructions
+        """
+        See also
+        --------
+        _transition_cancelled_fetch
+        _transition_cancelled_waiting
+        _transition_resumed_waiting
+        _transition_flight_fetch
+        """
+        if ts.previous == "flight":
+            if self.validate:
+                assert ts.coming_from
+                assert ts.next == "waiting"
+            if ts.done:
+                # We arrived here either from GatherDepNetworkFailureEvent or from
+                # GatherDepSuccessfulEvent but without the key in the data attribute.
+                # We would now normally try to fetch the task from another peer worker
+                # or transition it to missing if none are left; here instead we're going
+                # to compute the task as we had been asked by the scheduler.
+                ts.coming_from = None
+                ts.state = "released"
+                ts.done = False
+                ts.previous = None
+                ts.next = None
+                return {ts: "waiting"}, []
```

Comment:
I think we're missing a `in_flight_tasks.remove/discard(ts)` here

Reply:
I moved them all in `_gather_dep_done_common`